velox/functions/prestosql/ArrayCumSum.cpp

```diff
@@ -0,0 +1,175 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/RowsTranslationUtil.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+/// See documentation at
+/// https://prestodb.io/docs/current/functions/array.html#array_cum_sum
+///
+/// Implements the array_cum_sum function.
+///
+/// Zero element copy:
+///
+/// In order to prevent copies of array elements, the function reuses the
+/// internal elements() vector from the original ArrayVector to save the
+/// output ArrayVector.
+template <typename T>
+class ArrayCumSumFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    auto& arg = args[0];
+    VELOX_CHECK_EQ(
+        arg->encoding(),
+        VectorEncoding::Simple::ARRAY,
+        "Expected flat or constant encoding");
+    VectorPtr localResult;
+
+    if (arg->isConstantEncoding()) {
+      auto* constantArray = arg->as<ConstantVector<ComplexType>>();
+      const auto& flatArray = constantArray->valueVector();
+      const auto flatIndex = constantArray->index();
+
+      exec::LocalSingleRow singleRow(context, flatIndex);
+      localResult = applyFlat(*singleRow, flatArray, context);
+      localResult =
+          BaseVector::wrapInConstant(rows.end(), flatIndex, localResult);
+    } else {
+      localResult = applyFlat(rows, arg, context);
+    }
+
+    context.moveOrCopyResult(localResult, rows, result);
+  }
+
+ private:
+  VectorPtr applyFlat(
+      const SelectivityVector& rows,
+      const VectorPtr& arg,
+      exec::EvalCtx& context) const {
+    auto arrayVector = arg->as<ArrayVector>();
+    VELOX_CHECK(arrayVector);
+    memory::MemoryPool* pool = context.pool();
+
+    auto elementsVector = arrayVector->elements();
+    auto elementsRows =
+        toElementRows(elementsVector->size(), rows, arrayVector);
+    exec::LocalDecodedVector elements(context, *elementsVector, elementsRows);
+    auto rawElements = elementsVector->as<FlatVector<T>>()->mutableRawValues();
+    vector_size_t numRows = rows.end();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      auto size = arrayVector->sizeAt(row);
+      auto offset = arrayVector->offsetAt(row);
+      auto setNulls = [&](vector_size_t i) {
+        for (auto j = i; j < offset + size; j++) {
+          elementsVector->setNull(j, true);
+        }
+      };
+
+      if (elements->isNullAt(0)) {
```

Comment:
did you mean offset instead of 0 here? why not starting the loop bellow from offset

Reply:
Thanks for pointing it out, I have added it as a simple function as suggested.